File: solutions/ruby/all-your-base/1/all_your_base.rb
class BaseConverter
  def self.convert(input_base, digits, output_base)
    # 1. Validaciones de bases
    raise ArgumentError if input_base < 2 || output_base < 2
    
    # 2. Convertir de input_base a un número entero (Base 10)
    decimal_value = 0
    digits.reverse.each_with_index do |digit, index|
      raise ArgumentError if digit < 0 || digit >= input_base
      decimal_value += digit * (input_base**index)
    end

    # 3. Caso especial para el cero
    return [0] if decimal_value == 0

    # 4. Convertir de decimal a output_base
    result = []
    while decimal_value > 0
      # El residuo es el nuevo dígito en la base de destino
      result.unshift(decimal_value % output_base)
      # El cociente es lo que queda por procesar
      decimal_value /= output_base
    end

    result
  end
end

File: solutions/javascript/all-your-base/1/all-your-base.js
export const convert = (digits, inputBase, outputBase) => {
  // Validate bases
  if (inputBase < 2) {
    throw new Error('Wrong input base');
  }
  if (outputBase < 2) {
    throw new Error('Wrong output base');
  }
  
  // Validate digits array
  if (!Array.isArray(digits) || digits.length === 0) {
    throw new Error('Input has wrong format');
  }
  
  // Check for leading zeros
  if (digits.length > 1 && digits[0] === 0) {
    throw new Error('Input has wrong format');
  }
  
  // Validate each digit
  for (const digit of digits) {
    if (typeof digit !== 'number' || digit < 0 || digit >= inputBase) {
      throw new Error('Input has wrong format');
    }
  }
  
  // Handle zero
  if (digits.length === 1 && digits[0] === 0) {
    return [0];
  }
  
  // Convert from input base to decimal
  let decimalValue = 0;
  for (let i = 0; i < digits.length; i++) {
    decimalValue = decimalValue * inputBase + digits[i];
  }
  
  // Convert from decimal to output base
  const outputDigits = [];
  let remaining = decimalValue;
  
  while (remaining > 0) {
    outputDigits.unshift(remaining % outputBase);
    remaining = Math.floor(remaining / outputBase);
  }
  
  return outputDigits;
};